fix(devtools): ignore DOM Nodes from other frames when  performing render tree detection

AleksanderBodurri opened the pull request:
Previously, if an application had DOM Nodes injected into it from other frames, DevTools would fail to parse component trees with the render tree strategy properly because of an instanceof Node check that the framework performs.

Now we check for instanceof Node before even calling framework debug APIs on DOM nodes so that we can skip nodes that come from other frames entirely.

See #47357 for more context and a minimal reproduction of the issue.

pkozlowski-opensource reviewed (changes requested):
@AleksanderBodurri thnx for the PR - I guess it fixes the issue you've encountered while working on dev tools? What was the exact scenario?

Before we can merge this one we should add a test to illustrate the scenario where you've bumped into problems. Otherwise this code change without test might get removed during next round of refactorings and we will see the problem re-surfaced.

AleksanderBodurri commented:
@pkozlowski-opensource Added a test for this case, as well as some other tests for assertDomNode as I don't think they existed previously. The context of how this change relates DevTools can be found in #47357 along with a StackBlitz reproduction. Theres also this comment from the old DevTools repo that describes the issue more thoroughly https://github.com/rangle/angular-devtools/issues/910#issuecomment-905206933

pkozlowski-opensource commented on packages/core/src/util/assert.ts:
```diff
@@ -106,11 +106,21 @@ export function throwError(msg: string, actual?: any, expected?: any, comparison
       (comparison == null ? '' : ` [Expected=> ${expected} ${comparison} ${actual} <=Actual]`));
 }
 
+function assertDomNodeFromAnotherFrame(node: any) {
```
nit: this function doesn't assert but rather checks a condition. Should we rename to something like `isDomNodeFromAnotherFrame`?

pkozlowski-opensource commented on packages/core/src/util/assert.ts:
```diff
@@ -106,11 +106,21 @@ export function throwError(msg: string, actual?: any, expected?: any, comparison
 }
 
+function assertDomNodeFromAnotherFrame(node: any) {
+  if (node === null || typeof node !== 'object' || node instanceof Node) {
```
nit: I find it surprising that we are testing for `instanceof Node` here

pkozlowski-opensource commented:
@AleksanderBodurri after thinking about it a bit more I wonder if the best place for this check wouldn't be in the dev tools? With the proposed change LViews all of the sudden could have DOM nodes from different documents and this is not the setup we want to have in the normal use of Angular, right?

Let's chat offline to see if / how we could solve it differently, without modifying framework asserts.

AleksanderBodurri committed:
fix(devtools): ignore DOM Nodes from other frames when performing render tree detection

Previously, if an application had DOM Nodes injected into it from other frames, DevTools would fail to parse component trees with the render tree strategy properly because of an instanceof Node check that the framework performs.

Now we check for instanceof Node before even calling framework debug APIs on DOM nodes so that we can skip nodes that come from other frames entirely.
